src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1252,5 +1253,345 @@ internal bool CanAddParsedValue(HttpHeaderParser parser)
 
             internal bool IsEmpty => (RawValue == null) && (InvalidValue == null) && (ParsedValue == null);
         }
+
+
+        #region Low-level implementation details that work with _headerStore directly
+
+        // Used to store the CollectionsMarshal.GetValueRefOrAddDefault out parameter
+        private static bool s_dictionaryGetValueRefOrAddDefaultExistsDummy;
+
+        private const int InitialCapacity = 4;
+        internal const int ArrayThreshold = 64; // Above this threshold, header ordering will not be preserved
+
+        internal HeaderEntry[]? GetEntries()
+        {
+            object? store = _headerStore;
+            if (store is null)
+            {
+                return null;
+            }
+            else if (store is HeaderEntry[] entries)
+            {
+                return entries;
+            }
+            else
+            {
+                return GetEntriesFromDictionary(Unsafe.As<Dictionary<HeaderDescriptor, object>>(store));
+            }
+
+            static HeaderEntry[] GetEntriesFromDictionary(Dictionary<HeaderDescriptor, object> dictionary)
+            {
+                var entries = new HeaderEntry[dictionary.Count];
+                int i = 0;
+                foreach (KeyValuePair<HeaderDescriptor, object> entry in dictionary)
+                {
+                    entries[i++] = new HeaderEntry
+                    {
+                        Key = entry.Key,
+                        Value = entry.Value
+                    };
+                }
+                return entries;
+            }
+        }
+
+        internal HeaderEntry[]? GetEntriesGroupedByName()
+        {
+            GroupAllEntriesByName();
+            return GetEntries();
+        }
+
+        private bool EntriesAreLiveView => _headerStore is HeaderEntry[];
+
+        internal int Count
+        {
+            get
+            {
+                GroupAllEntriesByName();
+                return _count;
+            }
+        }
+
+        private void GroupAllEntriesByName()
+        {
+            if (_entriesMayBeUngrouped && _headerStore is HeaderEntry[] entries)
+            {
+                _entriesMayBeUngrouped = false;
+
+                for (int i = 0; i < entries.Length; i++)
+                {
+                    HeaderDescriptor key = entries[i].Key;
+                    if (!key.HasValue)
```

Comment:
Looking at all the places you check for key.HasValue, it seems like it's always used to indicate the end of the table when enumerating.

But it seems like in all these cases, we already have the _count. So why aren't we using _count to determine the end of the table in these cases?

For example, here you could just change to 

```C#
  for (int i = 0; i < _count; i++)
```

This seems like it would avoid the minor complication of allowing HeaderDescriptor to be uninitialized.

Reply:
> Also, it seems like there are several related changes here:
(1) Shrink HeaderDescriptor struct
(2) Rework header table

Previously these were more intertwined, hence I combined the changes.
In the current state of the PR, you are right, they can be made separately.

I'm updating the implementation now to take advantage of `_count` over `HasValue`.